50.003 Elements of Software Construction/Java Concurrency.md

```diff
@@ -0,0 +1,473 @@
+- Strive to use open calls (calling a method with no locks held) throughout your program.
+- A program will be free of lock-ordering deadlocks if all threads acquire the locks they need in a fixed global order.
+
+#### Non-blocking algorithms
+An algorithm is called non-blocking if failure or suspension of any thread cannot cause failure or suspension of another thread. 
+
+Non-blocking algorithms are immune to deadlock (though, in unlikely scenarios, may exhibit livelock or starvation)
+
+## Java Synchronizers
+### CountdownLatch
+The CountDownLatch class allows one or more threads to wait until a set of operations in other threads complete. CountDownLatch works by having a counter initialized with number of threads, which is decremented each time a thread complete its execution. When count reaches to zero, it means all threads have completed their execution, and thread waiting on latch resume the execution.
+
+1. `CountDownLatch latch = new CountDownLatch(numberOfThreads);
+`
+2. To inform the main thread that one of the service threads has completed its task, we decrement the counter with `latch.countDown();`
+3. The main thread can resume its work with `latch.await();`
+
+A more concrete example of a program that relies on service threads and a main thread needs to wait for several threads to finish their tasks:
+```
+public static void main (String[] args) {
+    CountDown Latch latch = new CountDownLatch(numberOfTasks);
+
+     // initiatite the tasks
+    services = new ArrayList<ToDoTask>();
+    for (int i = 0; i < numberOfTasks; i++) {
+        services.add(new ServiceWorker(i, latch));
+    }
+
+    // initiate the thread pool to take on the tasks
+    Executor executor = Executors.newFixedThreadPool(services.size());
+    for(ToDoTask w : services) {
+        executor.execute(w);
+    }
+
+    // Check
+    latch.await();
+    for(ToDoTask w: services) {
+        if(!w.isServiceUp()) {
+            return false;
+        }
+    }
+    
+    return true;
+    
+    }
+ }
+ ```
+ Here's the methods for ToDoTask object class.
+ ```
+ public ToDoTask(int i, CountDownLatch latch)
+    {
+        super();
+        this._latch = latch;
+        this._index = i;
+        this._serviceUp = false;
+    }
+ 
+    @Override
+    public void run() {
+        try {
+            doSomeWork();
+            _serviceUp = true;
+        } catch (Throwable t) {
+            t.printStackTrace(System.err);
+            _serviceUp = false;
+        } finally {
+            if(_latch != null) {
+                _latch.countDown(); // decrement latch counter.
+            }
+        }
+    }
+    
+    public doSomeWork() {
+        // some working code
+    }
+    
+```
+
+What if you are not certain of how many batches of tasks you actually need to complete? You would need to create a new CountDownLatch for every batch of tasks that you receive, and in this code, this is not accounted for. So, we turn to CyclicBarrier for resolving this issue.
```

Comment:
Wording can be more precise. Maybe something like:

Following the method shown by this code, a new `CountDownLatch` object would need to somehow be created for every batch of tasks if you are uncertain of the total number of batches. To address this problem in an elegant manner, we can use `CyclicBarrier` (next section)

Reply:
Great advice, thanks!